File: DSA/SakshamKashyap_590014183/day74/day74_Q2.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def is_balanced(root):
    def check(node):
        if not node:
            return 0, True
        lh, lb = check(node.left)
        rh, rb = check(node.right)
        balanced = lb and rb and abs(lh - rh) <= 1
        return 1 + max(lh, rh), balanced
    return check(root)[1]

# Example usage:
if __name__ == "__main__":
    # Tree: 1 / 2 / 3
    root1 = TreeNode(1, TreeNode(2, TreeNode(3)))
    print("Yes" if is_balanced(root1) else "No")  # Output: No

    # Tree: 1 / \ 2 3
    root2 = TreeNode(1, TreeNode(2), TreeNode(3))
    print("Yes" if is_balanced(root2) else "No")  # Output: Yes

File: DSA/IshikaGarg_500124973/Day_74_Q2.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

// Define a tree node
struct Node {
    int data;
    struct Node* left;
    struct Node* right;
};

// Create a new node
struct Node* newNode(int data) {
    struct Node* node = (struct Node*)malloc(sizeof(struct Node));
    node->data = data;
    node->left = node->right = NULL;
    return node;
}

// Utility function: returns height if balanced, -1 if not balanced
int checkHeight(struct Node* root) {
    if (root == NULL)
        return 0;

    int leftHeight = checkHeight(root->left);
    if (leftHeight == -1) return -1; // Left not balanced

    int rightHeight = checkHeight(root->right);
    if (rightHeight == -1) return -1; // Right not balanced

    if (abs(leftHeight - rightHeight) > 1)
        return -1; // Current node not balanced

    return (leftHeight > rightHeight ? leftHeight : rightHeight) + 1;
}

// Function to check if tree is balanced
bool isBalanced(struct Node* root) {
    return checkHeight(root) != -1;
}

// Driver code
int main() {
    /* Example 1:
            1
           /
          2
         /
        3
       Output: No
    */
    struct Node* root1 = newNode(1);
    root1->left = newNode(2);
    root1->left->left = newNode(3);

    if (isBalanced(root1))
        printf("Yes\n");
    else
        printf("No\n");

    /* Example 2:
            1
           / \
          2   3
       Output: Yes
    */
    struct Node* root2 = newNode(1);
    root2->left = newNode(2);
    root2->right = newNode(3);

    if (isBalanced(root2))
        printf("Yes\n");
    else
        printf("No\n");

    return 0;
}

